Move cache declarations to @apollo/cache-control-types (#6764)
apollographql/apollo-server: docs/source/migration.mdx

You can no longer type `CacheScope.Public` or `CacheScope.Private`. Instead, just use the string `'PUBLIC'` or `'PRIVATE'`. Values defined as `CacheScope` will only accept those two values, so any typos are still caught at compile time.

> `CacheScope` is now exported from the new `@apollo/cache-control-types` package rather than from an Apollo Server package directly. This enables libraries that work with multiple GraphQL servers (such as `@apollo/subgraph`) to refer to it without needing to depend on all of `@apollo/server`.

Author: Should be "You can now import" rather than export

> `CacheScope` is now exported from the new `@apollo/cache-control-types` package rather than from an Apollo Server package directly. This enables libraries that work with multiple GraphQL servers (such as `@apollo/subgraph`) to refer to it without needing to depend on all of `@apollo/server`.

Author: ```suggestion
You can now import `CacheScope` from the new `@apollo/cache-control-types` package (instead of importing it from an Apollo Server package). This enables libraries that work with multiple GraphQL servers (such as `@apollo/subgraph`) to refer to `CacheScope` without depending on `@apollo/server`.
```

Apollo Server 4 introduces `@apollo/utils.fetcher`, which defines a minimal fetch API (`Fetcher`) that provides Fetch API TypeScript typings. It is similar to `apollo-server-env` but has a clearer name and only supports argument structures that are likely to be compatible across many implementations of the Fetch API. (Specifically, it does not allow you to pass `Request` or `Headers` objects to `fetch`, because libraries often only know how to recognize their own implementations of these interfaces.)

### `@apollo/cache-control-types`

The `CacheHint`, `CacheAnnotation`, `CachePolicy`, and `ResolveInfoCacheControl` types are now exported from the new `@apollo/cache-control-types` package rather than from an Apollo Server package directly. This enables libraries that work with multiple GraphQL servers (such as `@apollo/subgraph`) to refer to it without needing to depend on all of `@apollo/server`. (This package also contains `CacheScope` which as described above is now a TypeScript-only symbol.)

> Apollo Server no longer uses the `declare module` TypeScript feature to declare that all `GraphQLResolveInfo` objects (ie, the `info` argument to resolvers) have a `cacheControl` field. Instead, `@apollo/cache-control-types` provides a `GraphQLResolveInfoWithCacheControl` interface that you can cast `info` to (if you don't want run-time validation), or functions `maybeCacheControlFromInfo` and `cacheControlFromInfo` if you do want runtime validation.

Reviewer: ```suggestion
Apollo Server 4 no longer uses the `declare module` TypeScript feature to declare that all `GraphQLResolveInfo` objects (i.e., the `info` argument to resolvers) have a `cacheControl` field. Instead, `@apollo/cache-control-types` provides a `GraphQLResolveInfoWithCacheControl` interface that you can cast `info` to (if you don't want run-time validation), or if you do want runtime validation, you can use the `maybeCacheControlFromInfo` and `cacheControlFromInfo` functions.
```

For example, if you had this resolver in Apollo Server 3:

```ts
  someField(parent, args, context, { cacheControl }) {
    cacheControl.setCacheHint({ maxAge: 100 });
  }
```

you can write this in Apollo Server 4:

```ts
import { cacheControlFromInfo } from '@apollo/cache-control-types';

// ...
  someField(parent, args, context, info) {
    cacheControlFromInfo(info).setCacheHint({ maxAge: 100 });
  }
```

> Alternatively, you can declare `info` to be of type `GraphQLResolveInfoWithCacheControl`; for example, if using `graphql-code-generator` with `typescript-resolvers`, you can use its `customResolveInfo` option.

Reviewer: ```suggestion
Alternatively, you can declare `info` to be of type `GraphQLResolveInfoWithCacheControl`. For example, if using `graphql-code-generator` with `typescript-resolvers`, you can use the `customResolveInfo` option.
```

> (This is a TypeScript-specific change; the runtime representation has not changed, and JavaScript code can continue to access `info.cacheControl` directly.)

Reviewer: ```suggestion
Note, this is a TypeScript-specific change, the runtime representation hasn't changed, and JavaScript code can continue to access `info.cacheControl` directly.
```

Author: Maybe

```suggestion
Note: this is a TypeScript-specific change. The runtime representation hasn't changed, and JavaScript code can continue to access `info.cacheControl` directly.
```
